apps/workspace-host/src/lib/socket-server.js

```diff
@@ -1,30 +1,68 @@
 // @ts-check
 const { Server } = require('socket.io');
-const redis = require('redis');
 const { createAdapter } = require('@socket.io/redis-adapter');
+const { Redis } = require('ioredis');
+const { logger } = require('@prairielearn/logger');
 const path = require('path');
 const debug = require('debug')('prairielearn:' + path.basename(__filename, '.js'));
 
 const { config } = require('./config');
 
+/**
+ * @param {Redis} client
+ * @param {string} type
+ */
+function attachEventListeners(client, type) {
+  client.on('error', (err) => {
+    logger.error(`redis client event for ${type}: error`, err);
+  });
+  client.on('connect', () => {
+    logger.verbose(`redis client event for ${type}: connect`);
+  });
+  client.on('ready', () => {
+    logger.verbose(`redis client event for ${type}: ready`);
+  });
+  client.on('reconnecting', (reconnectTimeMilliseconds) => {
+    logger.verbose(
+      `redis client event for ${type}: reconnecting in ${reconnectTimeMilliseconds} milliseconds`,
+    );
+  });
+  client.on('close', () => {
+    logger.verbose(`redis client event for ${type}: close`);
+  });
+  client.on('end', () => {
+    logger.verbose(`redis client event for ${type}: end`);
+  });
+  client.on('wait', () => {
+    logger.verbose(`redis client event for ${type}: wait`);
+  });
+  client.on('select', () => {
+    logger.verbose(`redis client event for ${type}: select`);
+  });
+}
+
 module.exports.init = async function (server) {
   debug('init(): creating socket server');
   module.exports.io = new Server(server);
   if (config.redisUrl) {
     // Use redis to mirror broadcasts via all servers
     debug('init(): initializing redis pub/sub clients');
-    module.exports.pub = redis.createClient({ url: config.redisUrl });
-    module.exports.sub = redis.createClient({ url: config.redisUrl });
-    await Promise.all([module.exports.pub.connect(), module.exports.sub.connect()]);
+    module.exports.pub = new Redis(config.redisUrl);
+    module.exports.sub = new Redis(config.redisUrl);
+
+    attachEventListeners(module.exports.pub, 'pub');
+    attachEventListeners(module.exports.sub, 'sub');
+
+    // TODO: remove the `sub` client if this issue is resolved:
+    // https://github.com/socketio/socket.io-redis-adapter/issues/516
```

Comment:
I opened this issue after realizing that we're wasting CPU + bandwidth receiving useless messages from Redis.

Reply:
I think this might be what https://github.com/socketio/socket.io-redis-emitter is for?

Reply:
Wow, great find!! I had no idea that existed haha. Let me move to that in this PR.

Reply:
I made this change in 7ed83f7 / ee338ab. @mwest1066 can you take another look at this PR now? I tested this locally and it seems to be working well.

Reply:
lgtm!